http/php/guzzle/src/Middleware/KiotaMiddleware.php

```diff
@@ -0,0 +1,38 @@
+<?php
+/**
+ * Copyright (c) Microsoft Corporation.  All Rights Reserved.
+ * Licensed under the MIT License.  See License in the project root
+ * for license information.
+ */
+
+
+namespace Microsoft\Kiota\Http\Middleware;
+
+use Microsoft\Kiota\Http\Middleware\Options\RetryOption;
+
+/**
+ * Class KiotaMiddleware
+ *
+ * Utility methods to expose middleware components and easily add them to Guzzle's HandlerStack
+ *
+ * @package Microsoft\Kiota\Http\Middleware
+ * @copyright 2021 Microsoft Corporation
+ * @license https://opensource.org/licenses/MIT MIT License
+ * @link https://developer.microsoft.com/graph
+ */
+class KiotaMiddleware
+{
+    /**
+     * Middleware that retries requests for 429,503 and 504 response status codes (by default) while respecting the Retry-After response header
+     * Configurable using {@link RetryOption}
+     *
+     * @param RetryOption|null $retryOption
+     * @return callable
+     */
+    public static function retry(?RetryOption $retryOption): callable
```

Comment:
would this be comparable to an extension method in .net?

Reply:
I'd say it's more of a Utility method to allow developers to easily add the retry handler to Guzzle's middleware pipeline aka HandlerStack.
Guzzle's pipeline is a stack of methods that accept the next handler to be called + the request & Guzzle config options
The methods here will allow a developer to easily add middleware e.g.
```php
$handlerStack = HandlerStack::create();
$retryOptions = new RetryOption();
$handlerStack->push(KiotaMiddleware::retry($retryOptions));
...
$guzzleClient = new \GuzzleHttp\Client(['handler' => $handlerStack]);
```
More context: [Guzzle docs](https://docs.guzzlephp.org/en/stable/handlers-and-middleware.html#middleware)
A similar pattern exposes built-in Guzzle middleware [here](https://github.com/guzzle/guzzle/blob/master/src/Middleware.php)

Reply:
so the client factory would basically build the stack for them? kind of like it builds an array for go/typescript?

Reply:
No, the dev would pass in the HandlerStack to the client factory. Opting for this approach to allow developers to fully control the pipeline i.e. add our middleware alongside other default Guzzle middleware/their custom middleware. So the client factory could have something like `setMiddleware(HandlerStack $handlerStack)`
Thoughts?

Reply:
it looks similar enough to what we have with other languages. How would one remove a handler from the stack? or even reorder it? I'm assuming the stack provide a set of APIs for that?

Reply:
Yes, the HandlerStack has APIs to add/remove/reorder handlers

Reply:
thanks for all the precisions, resolving